src/nr_openai_observability/langchain_callback.py

```diff
@@ -0,0 +1,205 @@
+        """Run when chain errors."""
+        tags = {error: str(error)}
+        span = self.spans_stack.pop()
+        assert span["attributes"]["name"] == "LlmChain"
+        self.finish_and_record_span(span, tags)
+
+    def on_tool_start(
+        self, serialized: Dict[str, Any], input_str: str, **kwargs: Any
+    ) -> Any:
+        """Run when tool starts running."""
+        tags = {
+            "tool_name": serialized.get("name"),
+            "tool_description": serialized.get("description"),
+            "tool_input": input_str,
+        }
+        self.spans_stack.append(self.create_span(name="LlmTool", tags=tags))
+
+    def on_tool_end(self, output: str, **kwargs: Any) -> Any:
+        """Run when tool ends running."""
+        tags = {
+            "tool_output": output,
+            "tool_invocation_counter": self.get_and_update_tool_invocation_counter(),
+        }
+        span = self.spans_stack.pop()
+
+        assert span["attributes"]["name"] == "LlmTool"
+        tool_name = kwargs.get("name")
+        assert span["attributes"]["tool_name"] == tool_name
+
+        self.finish_and_record_span(span, tags)
+
+    def on_tool_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> Any:
+        """Run when tool errors."""
+        tags = {
+            "error": error,
+        }
+        span = self.spans_stack.pop()
+        assert span["attributes"]["name"] == "LlmTool"
+        tool_name = kwargs.get("name")
+        assert span["attributes"]["tool_name"] == tool_name
+
+        self.finish_and_record_span(span, tags)
+
+    def on_text(self, text: str, **kwargs: Any) -> Any:
+        """Run on arbitrary text."""
+
+    def on_agent_action(self, action: AgentAction, **kwargs: Any) -> Any:
+        """Run on agent action."""
+
+    def on_agent_finish(self, finish: AgentFinish, **kwargs: Any) -> Any:
+        """Run on agent end."""
+
+    def parent_id_callback(self) -> str:
+        parent_span = self.spans_stack[-1] if self.spans_stack else None
+        return parent_span["id"] if parent_span else None
+
+    def create_span(
+        self,
+        name: Optional[str] = None,
+        tags: Optional[Dict[str, Any]] = None,
+        guid: Optional[str] = None,
+        trace_id: Optional[str] = None,
+        parent_id: Optional[str] = None,
+        start_time_ms: Optional[int] = None,
+        duration_ms: Optional[int] = None,
+    ):
+        if parent_id is None:
+            parent_span = self.spans_stack[-1] if self.spans_stack else None
+            parent_id = parent_span["id"] if parent_span else None
+
+        if self.langchain_callback_metadata:
+            tags = tags or {}
+            tags.update(self.langchain_callback_metadata)
+
+        if not trace_id and "newrelic" in sys.modules:
+            import newrelic.agent
+
+            trace_id = getattr(newrelic.agent.current_transaction(), 'trace_id', None)
+
+        trace_id = trace_id or self.trace_id
```

Comment:
I think we run into an issue here if the user is not using the language agent and tries to re-use the callback handler (since the `self.trace_id` is only defined once). Should we explicitly call out in the readme / example that the user should construct the callback handler once for each request?